### PR TITLE
Feature: Diagonal gradients on rectangles

### PR DESCRIPTION
**Pull request by Gemba**
Refers to https://retropie.org.uk/forum/topic/30294/the-gradient-experience-progressbar

### CODE REVIEW

**Comment by pjft** [comment]
Hi. So this is an area I am not at all familiar with from ES, so if someone else wants to chime in please do.

Can you elaborate a bit more on the logic and behavior expected, from a user and technical standpoint? I read the forum post, as well as the code, but one thing I'm (initially) confused about is, if DEFAULT is "// diagonal towards endcolor in bottom-right", isn't that supposed to be 0x03 - as it's TO_BOTTOM and TO_RIGHT together? Or am I missing something in the logic, as I don't see it being used anywhere else in the changes (unless it already was)?

Also - and here comes my lack of knowledge from this part of the code - where/how does the user set those other gradient types? I imagine THEMES.md should be updated? Or is that not present yet - and, as such, the current other enums are just there for future usage as needed?

Please provide some mode details here, and @jrassa if you're around and want to chime in as you have a lot more context on this, please do.

Thanks, and if I'm completely wrong - which might be the case at 8AM - feel free to say so.

Have a great day!

**Comment by jrassa** [comment]
> Can you elaborate a bit more on the logic and behavior expected, from a user and technical standpoint? I read the forum post, as well as the code, but one thing I'm (initially) confused about is, if DEFAULT is "// diagonal towards endcolor in bottom-right", isn't that supposed to be 0x03 - as it's TO_BOTTOM and TO_RIGHT together? Or am I missing something in the logic, as I don't see it being used anywhere else in the changes (unless it already was)?

Agreed.   Although I would recommend just dropping the `DEFAULT` value.

> Also - and here comes my lack of knowledge from this part of the code - where/how does the user set those other gradient types? I imagine THEMES.md should be updated? Or is that not present yet - and, as such, the current other enums are just there for future usage as needed?

From my reading, this change does not expose the option for diagonal gradients to themes, it just adds support to the Renderer.

File: es-core/src/renderers/Renderer.h [hunk 1]
```diff
@@ -29,6 +30,19 @@ namespace Renderer
...
+	namespace Gradient
+	{
+		enum Mode : uint8_t
+		{
+			NONE        = 0x00, // off, no gradient
+			TO_BOTTOM   = 0x01, // straight along y axis towards endcolor
+			TO_RIGHT    = 0x02, // straight along x axis towards endcolor
+			DEFAULT     = 0x04, // diagonal towards endcolor in bottom-right
+			FLIP_LR     = 0x08, // (diagonal) towards endcolor in left
+			FLIP_TOPBTM = 0x10  // (diagonal) towards endcolor in top
```
Comment (jrassa) [inline es-core/src/renderers/Renderer.h:42]: Shouldn't this also be a power of 2?  As is, combining `TO_RIGHT` and `FLIP_LR`, would also match `FLIP_TOPBTM`

File: es-core/src/renderers/Renderer.h [hunk 2]
```diff
@@ -67,6 +81,7 @@ namespace Renderer
 	void        pushClipRect    (const Vector2i& _pos, const Vector2i& _size);
 	void        popClipRect     ();
 	void        drawRect        (const float _x, const float _y, const float _w, const float _h, const unsigned int _color, const unsigned int _colorEnd, bool horizontalGradient = false, const Blend::Factor _srcBlendFactor = Blend::SRC_ALPHA, const Blend::Factor _dstBlendFactor = Blend::ONE_MINUS_SRC_ALPHA);
+	void        drawRect        (const float _x, const float _y, const float _w, const float _h, Gradient::Mode _mode, const unsigned int _color, const unsigned int _colorEnd, const Blend::Factor _srcBlendFactor = Blend::SRC_ALPHA, const Blend::Factor _dstBlendFactor = Blend::ONE_MINUS_SRC_ALPHA);
```
Comment (jrassa) [inline es-core/src/renderers/Renderer.h:84]: Any reason why the argument order was changed? In the existing call, the gradient parameter came after the colors, but in this new method it comes before.  It would be good to have consistency here.  It would make it easier to eventually migrate away from the old method.

I would have expected something along the lines of the following, with gradient mode after the colors, and a default value for gradient mode matching the behavior of the current `drawRect`:  

```c++
void drawRect(const float _x, const float _y, const float _w, const float _h, const unsigned int _color, const unsigned int _colorEnd, Gradient::Mode _mode = Gradient::Mode::TO_BOTTOM, const Blend::Factor _srcBlendFactor = Blend::SRC_ALPHA, const Blend::Factor _dstBlendFactor = Blend::ONE_MINUS_SRC_ALPHA);
```

File: es-core/src/renderers/Renderer.cpp [hunk 3]
```diff
@@ -254,17 +254,35 @@ namespace Renderer
...
+		if (_mode & Gradient::FLIP_LR) {
+			for(int i = 0; i < 4; ++i)  {
+				vertices[i].pos[0] = vertices[i].pos[0] + ((vertices[i].pos[0] == _x) ? _w : -_w);
+			}
+		}
+		if (_mode & Gradient::FLIP_TOPBTM) {
+			for(int i = 0; i < 4; ++i)  {
+				vertices[i].pos[1] = vertices[i].pos[1] + ((vertices[i].pos[1] == _y) ? _h : -_h);
+			}
+		}
```
Comment (jrassa) [inline es-core/src/renderers/Renderer.cpp:284]: Would it be easier to swap the colors vs modifying the vertices?

File: es-core/src/renderers/Renderer.h (hunk 1 above)
Comment (Gemba) [reply es-core/src/renderers/Renderer.h:42]: The joy of hex: `0x10 == 0x08<<1`

File: es-core/src/renderers/Renderer.h (hunk 2 above)
Comment (Gemba) [reply es-core/src/renderers/Renderer.h:84]: Can be considered, if the gradient parameter is made mandatory for both methods. Otherwise autocasting can have undesired results.

File: es-core/src/renderers/Renderer.cpp (hunk 3 above)
Comment (Gemba) [reply es-core/src/renderers/Renderer.cpp:284]: I started this way but it turned out to be too hard to read/understand the code and errorprone to maintain. This implementation is easier to maintain.

**Comment by Gemba** [comment]
Thanks for challenging my PR. 
I assume I was too terse on the posting and here. So I will give some more information about the implementation.

Currently this gradient is not available theme-wise, it can be done in a second PR after we agreed on this one. (As I am writing this I have the idea of allowing an individual color for each corner...)

 `DEFAULT` is a fall through value which does not change anything thus one gets a gradient from top-left to bottom-right running along the line segment top-left to bottom-right. One can set one start color (`color`) and one end color (`endColor`), like this:

Grad_default

However, `DEFAULT` and `TO_RIGHT`+`TO_BOTTOM` do not yield the same result. 

If you set `TO_RIGHT` the top-right corner will be set to `endColor` resulting in a gradient running from left to right along the x-axis.
If you set `TO_BOTTOM` the bottom-left corner will be set to `endColor` resulting in a gradient running from top to bottom along the y-axis. Combined it will look like this:

Grad_to_right_and_to_bottom

If one applies the flip options along both axis's and swap also the values of `color` and `endColor` the result is identical to the one shown as `DEFAULT`:

Grad_To_Right_and_Bottom_plus_colors_swapped


Hope this sheds some light on the implementation.

File: es-core/src/renderers/Renderer.h [hunk 4]
```diff
@@ -29,6 +30,19 @@ namespace Renderer
...
 	} // Blend::
 
+	namespace Gradient
+	{
+		enum Mode : uint8_t
+		{
+			NONE        = 0x00, // off, no gradient
+			TO_BOTTOM   = 0x01, // straight along y axis towards endcolor
+			TO_RIGHT    = 0x02, // straight along x axis towards endcolor
+			DEFAULT     = 0x04, // diagonal towards endcolor in bottom-right
```
Comment (pjft) [inline es-core/src/renderers/Renderer.h:40]: @Gemba I thought I had asked about this, but I might not have sent it. Shouldn't default be 0x03, if it's TO_BOTTOM and TO_RIGHT at the same time?

Comment (Gemba) [reply es-core/src/renderers/Renderer.h:40]: Ah, I remember @pjft  but it escaped my mind. Maybe the constantnames are confusing. `TO_RIGHT` means running the gradient along x-axis only (top-right and button-right will be set to `endColor`). `TO_BOTTOM` means running the gradient along y-axis only (bottom-left and bottom-right will be set to `endColor`). In contrast `DEFAULT` sets only bottom-right to `endColor` resulting in an gradient running diagonal to screen orientation along the axis top-left to bottom-right. If you can think of more precise constants: I am open to suggestions. This is also reflected in the first two pictures I provided here. First is `DEFAULT`, second is both `TO_RIGHT` and `TO_BOTTOM` set.

Comment (pjft) [reply es-core/src/renderers/Renderer.h:40]: Actually, there are other options for the diagonal gradient, but this possibly one of them.

**Comment by Gemba** [comment]
I dug a deeper into this issue (but it is not a rabbit-hole after all). I came to the conclusion it would be more elegant to do shaders with GLSL, the current implementation uses triangularization (hence the 4 vectors to span two triangles). This also is the reason that  one would have  to calculate the colors for the dangling corners (as you outlined @pjft ) or have the user to provide it.

GLSL would have -among others- these advantages: Offloading to GPU, proven solution to be working (less bugs). 
I am leaning towards linear gradients only (which would be the major use case for theming), with a similar wording and semantic as in CSS linear gradients.

TL;DR: If "linear gradients in themes" is on the project's roadmap (on the cool-features-map), I could come up with at least the core implementation (theme gradient via XML would be a second PR). If not I will retreat this current PR.

**Comment by tomaz82** [comment]
@Gemba please explain further what shaders would do in this case.

**Comment by Gemba** [comment]
To my knowledge the (horizontal) GradientType can be applied (theme-wise) to the SystemView*s carousel, any ImageComponent and TextListComponent. These could then use linear gradients (shaders) defined via theme XML. Maybe there are more usecases.
However, if this is YAGNI or not envisioned, then -as I stated earlier- I can retreat this suggested change. It was fun digging into this topic nevertheless.

**Comment by pjft** [comment]
@tomaz82 @jrassa any further comments here?

**Comment by Gemba** [comment]
I will withdraw this PR as it seems it is not of broad interest.
